### PR TITLE
Ensure GDExtension class is the correct type for the Godot engine class

### LINKED ISSUE #995: `Object::cast_to` works incorrectly for GDExtension custom classes
Copy of [this issue#71332](https://github.com/godotengine/godot/issues/71332#issue-1532399391) from main repository, as it is related to GDExtension.

### Issue description

`Object::cast_to` does not work properly in case of custom GDExtension classes. For example, a `nullptr` is NOT returned in some cases where an object can not be casted to the desired type:

(Let `TestNode` inherit from `Node2D`, `some_node_2d` be of type `Node2D`)
```cpp
godot::Node::cast_to<TestNode>(some_node_2d) != nullptr;  // true, must be false, as Node2D can not be casted to TestNode.
```

From my experience, it seems that `cast_to` works with custom classes as if they are their base classes, as (in this example) I can cast all the classes inherited from `Node2D` to `TestNode`, and can not cast any other type (see steps to reproduce and reproduction project).

The GDScript, although, works fine with casting of custom classes, see example. C++ code for casting in-engine types also seems to work fine.

Steps to reproduce and example project with GDExtension source code can be seen [here (issue#71332)](https://github.com/godotengine/godot/issues/71332#issue-1532399391)

### PR DESCRIPTION
**Pull request by dsnopek**
As discovered after merging PR #1037 (which was later reverted), when a Godot engine class is created on the Godot side and then passed to GDExtension, the GDExtension class that is created may not be of the correct type (most often it'll be created as `godot::Object` and then be unsafely cast to the correct class, which could lead to strange/incorrect behavior later).

This is explained in detail in these two comments:

- https://github.com/godotengine/godot-cpp/issues/1042#issuecomment-1432251538
- https://github.com/godotengine/godot-cpp/issues/995#issuecomment-1434345271

This PR aims to ensure that the GDExtension class that's created to wrap a Godot engine class always matches the Godot class it wraps.

This is a draft currently, because there is still more that I'd like to do:

- ~~This isn't as efficient as it could be. We're always looking up the instance binding callbacks, even when we don't need them (which would be the case for any class created on the GDExtension side)~~
- ~~I'd like to make a utility method somewhere to correctly convert from a `GodotObject*` to a `godot::Object`, rather than have it only exist in `Variant::operator Object*()`~~
- ~~As pointed out by @zhehangd in https://github.com/godotengine/godot-cpp/issues/995#issuecomment-1434345271, there are more places in godot-cpp that need to be fixed (for which we'll use the utility function)~~

This depends on PR https://github.com/godotengine/godot/pull/73511 to the Godot engine, because a new function needs to be added to the GDExtension interface.

~~UPDATE: Back in draft, since I'm rebasing this on https://github.com/godotengine/godot/pull/76406~~

UPDATE: https://github.com/godotengine/godot/pull/76406 was merged, so I'm taking this out of draft!

Fixes #995

### CODE REVIEW

**Comment by dsnopek** [comment]
My latest update does the following:

- Adds a "fast path"  for getting instance bindings when an object already has them. This way we aren't trying to look up the instance binding callbacks when we don't actually need them.
- Creates a `godot::internal::get_object_instance_binding(GodotObject *p_engine_object)` function that can be used to convert any Godot engine object into the correct godot-cpp object class, where the Godot engine class isn't known at compile time.
- Switched to the new function everywhere in godot-cpp that it was needed. I left alone the use of `internal::gde_interface->object_get_instance_binding()` when making the singleton objects, since the class is 100% at compile time.

So, once the CI has passed, I'm going to take this out of Draft!

(NOTE: just to make this absolutely clear, we don't need this for 4.0 stable - this can wait for 4.x :-))

**Comment by dsnopek** [comment]
Converting to a draft while PR https://github.com/godotengine/godot/pull/76406 is in progress

**Comment by dsnopek** [comment]
I've updated this to be based on https://github.com/godotengine/godot/pull/76406

So, this'll need to wait until at least that one is merged before this can come out of draft again.

However, I'd also like to add a test (based on PR #1101) to validate the fix, so we can't later regress!

So, this'll also need to wait for that PR to be merged.

**Comment by dsnopek** [comment]
This is finally really ready for review! I've also put it on the agenda for the next GDExtension meeting :-)

**Comment by dsnopek** [comment]
Discussed at the GDExtension meeting, and overall the solution makes sense. @vnen is going to do code review

**Review by vnen — APPROVED** [review]

**Comment by BrianWiz** [comment]
Hey do you need any help here? I hate to be that guy but that bug was unacceptable. Are there unit tests for this? If not I could look into writing some.

**Comment by Darker** [comment]
Thanks so much for this, I was waiting for it! Can't wait to update and try it :)

**Pull request merged**
